File: lib/validate.js
import { existsSync } from "node:fs";
import path from "node:path";
import { CONFIG_FILE, GROUPS, SETTINGS_FILE } from "./constants.js";
import { commandId, getCommands } from "./store.js";
import { color } from "./theme.js";

export function validateConfig(config, settings) {
  const checks = [];
  const commands = getCommands(config);
  const ids = commands.map(commandId);
  const duplicates = ids.filter((id, index) => ids.indexOf(id) !== index);
  const groups = new Set(Object.keys(GROUPS));

  checks.push({ ok: Array.isArray(config.commands), label: `${CONFIG_FILE} contains a commands array` });
  checks.push({ ok: duplicates.length === 0, label: "command ids are unique", detail: [...new Set(duplicates)].join(", ") });
  checks.push({ ok: commands.every((command) => command.name && command.command), label: "each command has name and command" });
  checks.push({ ok: commands.every((command) => groups.has(command.group)), label: "all command groups are known" });
  checks.push({ ok: Number.isInteger(settings.maxFavoriteCommands) && settings.maxFavoriteCommands >= 1, label: "maxFavoriteCommands is a positive integer" });
  checks.push({ ok: (settings.favoriteCommands ?? []).length <= settings.maxFavoriteCommands, label: "favoriteCommands respects maxFavoriteCommands" });
  checks.push({ ok: existsSync(path.join(process.cwd(), SETTINGS_FILE)) || true, label: `${SETTINGS_FILE} is optional and local` });

  return { ok: checks.every((check) => check.ok), checks, commandCount: commands.length };
}

export function printValidationReport(result) {
  console.log(color.bold("Validate kmc project"));
  console.log(color.dim("Checking config shape, command ids, groups, and local settings."));
  console.log("");
  for (const check of result.checks) {
    const marker = check.ok ? color.green("OK") : color.red("FAIL");
    console.log(`${marker} ${check.label}${check.detail ? color.dim(` (${check.detail})`) : ""}`);
  }
  console.log("");
  console.log(result.ok ? color.green("Everything looks okay.") : color.red("Validation failed."));
  console.log(`${result.commandCount} command${result.commandCount === 1 ? "" : "s"} configured.`);
}

File: internal/config/validate.go
package config

type Check struct {
	OK     bool
	Label  string
	Detail string
}

type ValidationResult struct {
	OK           bool
	Checks       []Check
	CommandCount int
}

func Validate(input Config, settings Settings) ValidationResult {
	checks := []Check{}
	commands, commandErr := Commands(input)
	groupIDs := make([]string, 0, len(input.Groups))
	for _, group := range input.Groups {
		groupIDs = append(groupIDs, group.ID)
	}
	duplicateGroups := duplicates(groupIDs)
	ids := make([]string, 0, len(commands))
	for _, command := range commands {
		ids = append(ids, CommandID(command))
	}
	duplicateCommands := duplicates(ids)
	known := map[string]bool{}
	for id := range KnownGroups {
		known[id] = true
	}
	for _, id := range groupIDs {
		known[id] = true
	}
	checks = append(checks,
		Check{OK: input.Groups != nil, Label: ConfigFile + " contains a groups array"},
		Check{OK: len(duplicateGroups) == 0, Label: "group ids are unique", Detail: join(duplicateGroups)},
		Check{OK: everyGroupComplete(input.Groups), Label: "each group has id, label, and type"},
		Check{OK: commandErr == nil && len(duplicateCommands) == 0, Label: "command ids are unique", Detail: join(duplicateCommands)},
		Check{OK: commandErr == nil && everyCommandComplete(commands), Label: "each command has name and command"},
		Check{OK: commandErr == nil && everyCommandGroupKnown(commands, known), Label: "all command groups are known"},
		Check{OK: everyKnown(settings.FavoriteGroups, known), Label: "favoriteGroups point to known groups"},
		Check{OK: everyKnown(settings.HiddenGroups, known), Label: "hiddenGroups point to known groups"},
		Check{OK: settings.MaxFavoriteCommands >= 1, Label: "maxFavoriteCommands is a positive integer"},
		Check{OK: len(settings.FavoriteCommands) <= settings.MaxFavoriteCommands, Label: "favoriteCommands respects maxFavoriteCommands"},
		Check{OK: true, Label: SettingsFile + " is optional and local"},
	)
	result := ValidationResult{OK: true, Checks: checks, CommandCount: len(commands)}
	for _, check := range checks {
		result.OK = result.OK && check.OK
	}
	return result
}

func duplicates(values []string) []string {
	seen, added := map[string]bool{}, map[string]bool{}
	var result []string
	for _, value := range values {
		if seen[value] && !added[value] {
			result = append(result, value)
			added[value] = true
		}
		seen[value] = true
	}
	return result
}

func join(values []string) string {
	var result string
	for i, value := range values {
		if i > 0 {
			result += ", "
		}
		result += value
	}
	return result
}

func everyGroupComplete(groups []Group) bool {
	for _, group := range groups {
		if group.ID == "" || group.Label == "" || group.Type == "" {
			return false
		}
	}
	return true
}

func everyCommandComplete(commands []Command) bool {
	for _, command := range commands {
		if command.Name == "" || command.Command == "" {
			return false
		}
	}
	return true
}

func everyCommandGroupKnown(commands []Command, known map[string]bool) bool {
	for _, command := range commands {
		if !known[command.GroupID] {
			return false
		}
	}
	return true
}

func everyKnown(values []string, known map[string]bool) bool {
	for _, value := range values {
		if !known[value] {
			return false
		}
	}
	return true
}
